### DISCUSSION TITLE
Internal constructor in BankIdLauncherCustomAppCallbackContext

### DISCUSSION THREAD

**Discussion by Max-Eriksson** [ActiveLogin/ActiveLogin.Authentication#428] [General]
To support native apps we're using the features [custom browser detection and launch info](https://github.com/ActiveLogin/ActiveLogin.Authentication/blob/main/docs/articles/bankid.md#custom-browser-detection-and-launch-info) and [custom app callback](https://github.com/ActiveLogin/ActiveLogin.Authentication/blob/main/docs/articles/bankid.md#shorthand-for-only-overriding-return-url-for-custom-apps).

However, the [constructor](https://github.com/ActiveLogin/ActiveLogin.Authentication/blob/25f0f8b446d79af196cccbbf527e19189ad32f69/src/ActiveLogin.Authentication.BankId.Core/Launcher/BankIdLauncherCustomAppCallbackContext.cs#L7) in BankIdLauncherCustomAppCallbackContext is internal and we can't use it when implementing the IBankIdLauncher interface, i.e I can't create a callback context as it's done in the [source](https://github.com/ActiveLogin/ActiveLogin.Authentication/blob/25f0f8b446d79af196cccbbf527e19189ad32f69/src/ActiveLogin.Authentication.BankId.Core/Launcher/BankIdLauncher.cs#L126).

Is this by design? If yes; How do I create a BankIdLauncherCustomAppCallbackContext?

BR,
Max.

**Comment by PeterOrneholm** [comment]
Good catch. We usually try to expose as little as possible, but I think we should make this constructor public. Will fix.

**Comment by Max-Eriksson** [comment]
Until a public constructor is available we're using the following workaround.

```C#
            ConstructorInfo info = typeof(BankIdLauncherCustomAppCallbackContext).GetTypeInfo().DeclaredConstructors.First();

            if (info.Invoke(new object[] { device, request }) is BankIdLauncherCustomAppCallbackContext customAppCallbackContext)
            {
                var customAppCallback = await GetRelevantCustomAppCallbackAsync(customAppCallbackContext, customAppCallbacks);

                if (customAppCallback != null)
                {
                    return await customAppCallback.GetCustomAppReturnUrl(customAppCallbackContext);
                }
            }
```

**Comment by PeterOrneholm** [comment]
@Max-Eriksson We have made this public now (https://github.com/ActiveLogin/ActiveLogin.Authentication/pull/434/files#diff-3e365192149746040d1f6b9e9641bb45031536d7505c194e8d8d47eb9f439e39). It's part of the 8.0.0-beta-1 release available on Nuget.
